### PR TITLE
expression, util: improve cast function signatures

### PR DESCRIPTION
**Pull request by XuHuaiyu**
Before this commit, the implementation of `castXXAsString` and `castXXAsDecimal` function signatures are not correct,
castXXAsString does not use the `flen`, `charset` and `Tp` conditions of the target FieldType,
castXXAsDecimal does not use the `flen` and `decimal` conditions of the target FieldType.

PTAL @coocood @shenli @hanfei1991

### CODE REVIEW

**Comment by hanfei1991** [comment]
I think we can add the `fieldType` attribute for baseBuiltinFunction.

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Comment by XuHuaiyu** [comment]
PTAL @coocood

File: expression/builtin_cast_test.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 	"github.com/pingcap/tidb/util/charset"
 	"github.com/pingcap/tidb/util/testleak"
 	"github.com/pingcap/tidb/util/types"
+	"strconv"
```
Comment (hanfei1991) [inline expression/builtin_cast_test.go]: format the imports

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @shenli

**Comment by XuHuaiyu** [comment]
cast as Datetime
cast as Duration
should consider fsp.

PTAL @shenli @hanfei1991

**Comment by shenli** [comment]
@XuHuaiyu Please fix CI.

**Comment by XuHuaiyu** [comment]
fixed @shenli

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**